File: backend-node/src/services/AgentService.js
const WeatherService = require('./WeatherService');

class AgentService {
    constructor() {
        this.isSummoning = false;
        this.timer = null;
        this.checkInterval = 60 * 1000; // Check every 1 minute
    }

    start() {
        console.log("[Agent] The Tyrant is watching...");
        // Start the loop
        this.timer = setInterval(() => this.tick(), this.checkInterval);
    }

    stop() {
        if (this.timer) clearInterval(this.timer);
    }

    stopSummon() {
        if (this.isSummoning) {
            console.log("[Agent] Summon stopped by User Action (NFC).");
            this.isSummoning = false;
        }
    }

    manualSummon() {
        console.log("[Agent] Manual Summon Triggered.");
        this.isSummoning = true;
        // Auto-stop after 5 mins if ignored? Or be relentless? 
        // For now, let's keep the manual summon relentless until checked or 5 mins.
        setTimeout(() => {
            if (this.isSummoning) {
                console.log("[Agent] Giving up manually after 5 mins.");
                this.isSummoning = false;
            }
        }, 5 * 60 * 1000);
    }

    async tick() {
        if (this.isSummoning) return; // Already screaming

        const now = new Date();
        const hour = now.getHours();

        // 1. Base Probability by Hour (The Tyrant's Schedule)
        let prob = 0.0;

        if (hour < 18) {
            prob = 0.0; // Day time: Peace
        } else if (hour < 20) {
            prob = 0.05; // 18:00-20:00: 5% / min (Very low, ~25% chance in 2 hours)
        } else if (hour < 22) {
            prob = 0.15; // 20:00-22:00: 15% / min (High, ~60% chance/hour)
        } else if (hour < 24) {
            prob = 0.40; // 22:00-00:00: 40% / min (Very High)
        } else {
            prob = 1.0;  // 00:00+: 100% (IMMEDIATE)
        }

        // 2. Weather Multiplier
        try {
            const temp = await WeatherService.getCurrentTemperature();
            if (temp !== null && temp < 10) {
                prob = prob * 1.5; // Cold? Get in now.
                console.log(`[Agent] It's cold (${temp}C). Patience reduced.`);
            }
        } catch (e) {
            // Ignore weather error
        }

        // 3. Roll the Dice
        const roll = Math.random();
        const shouldSummon = roll < prob;

        console.log(`[Agent] Tick ${hour}:00. Roll: ${roll.toFixed(2)} < Prob: ${prob.toFixed(2)} ? ${shouldSummon}`);

        if (shouldSummon) {
            console.log("[Agent] DECISION: SUMMON USER NOW.");
            this.isSummoning = true;
        }
    }
}

module.exports = new AgentService();

File: backend-rails/app/services/agent_service.rb
class AgentService
  include Singleton

  attr_accessor :is_summoning, :current_shame_message

  def initialize
    @is_summoning = false
    @current_shame_message = "I am ignoring my bath to play on my phone. #IntentlessBath"
    @check_interval = 60 # seconds
    @thread = nil
  end

  def start
    return if @thread&.alive?
    puts "[Agent] The Tyrant is watching..."
    @thread = Thread.new do
      loop do
        tick
        sleep @check_interval
      end
    end
  end

  def stop
    @thread&.kill
  end

  def stop_summon
    if @is_summoning
      puts "[Agent] Summon stopped by User Action (NFC)."
      @is_summoning = false
    end
  end

  def manual_summon
    puts "[Agent] Manual Summon Triggered."
    @is_summoning = true
    # Auto-stop logic in a separate thread/timer if needed
    Thread.new do
      sleep(5 * 60)
      if @is_summoning
        puts "[Agent] Giving up manually after 5 mins."
        @is_summoning = false
      end
    end
  end

  def tick
    return if @is_summoning

    now = Time.now
    hour = now.hour
    prob = 0.0

    if hour < 18
      prob = 0.0
    elsif hour < 20
      prob = 0.05
    elsif hour < 22
      prob = 0.15
    elsif hour < 24
      prob = 0.40
    else
      prob = 1.0
    end

    # Weather Multiplier
    temp = WeatherService.get_current_temperature
    if temp
      if temp < 10
        prob *= 2.0
        puts "[Agent] Cold boost."
      elsif temp > 30
        prob *= 1.5
        puts "[Agent] Hot boost."
      end
    end

    roll = rand
    should_summon = roll < prob
    
    puts "[Agent] Tick #{hour}:00. Roll: #{roll.round(2)} < Prob: #{prob.round(2)} ? #{should_summon}"

    if should_summon
      puts "[Agent] DECISION: SUMMON USER NOW."
      @is_summoning = true
      
      # Generate Dynamic Shame
      if temp && temp < 10
        @current_shame_message = "It is freezing (#{temp}C) and I refuse to warm up. I choose to suffer. #IntentlessBath"
      elsif hour >= 23
        @current_shame_message = "It is past 11PM and I am doomscrolling instead of bathing. Help me. #IntentlessBath"
      else
        @current_shame_message = "I am ignoring the Tyrant. My hygiene is optional. #IntentlessBath"
      end
    end
  rescue => e
    puts "[Agent] Error in tick: #{e.message}"
  end
end
